xcube/webapi/datasets/context.py

```diff
@@ -842,18 +818,52 @@ def _is_not_empty(prefix):
     return prefix != '' and prefix != '/' and prefix != '\\'
 
 
+def _get_roots(paths):
+    list.sort(paths)
+    prefixes = _get_common_prefixes(paths)
+    if len(prefixes) < 1:
+        roots = ['']
+    else:
+        # perform further step to merge prefixes with same start
+        prefixes = list(set(prefixes))
+        prefixes.sort()
+        roots = []
+        root_candidate = prefixes[0]
+        for root in prefixes[1:]:
+            common_root = os.path.commonprefix([root_candidate, root])
+            if _is_not_empty(common_root):
+                root_candidate = common_root
+            else:
+                roots.append(root_candidate)
+                root_candidate = root
+        roots.append(root_candidate)
+    if roots[0] == '':
+        roots.append(roots.pop(0))
+    return roots
+
+
 def _get_common_prefixes(p):
     # Recursively examine a list of paths for common prefixes:
     # If no common prefix is found, split the list in half and
     # examine each half separately
     prefix = os.path.commonprefix(p)
+    # ensure prefix does not end with full or partial directory
+    # or file name
+    prefix = prefix[:max(_lastindex(prefix, '/'), _lastindex(prefix, '\\'), 0)]
     if _is_not_empty(prefix) or len(p) == 1:
         return [prefix]
     else:
         return _get_common_prefixes(p[:int(len(p) / 2)]) + \
                _get_common_prefixes(p[int(len(p) / 2):])
 
 
+def _lastindex(prefix, symbol):
+    try:
+        return prefix.rindex(symbol)
+    except ValueError:
+        return -1
+
+
```

Comment:
Consider replacing the while function by `prefix.rfind(symbol)`. (https://docs.python.org/3/library/stdtypes.html#str.rfind)

Reply:
You cannot ask for a slice:

```
>>> ''[0]
''
```

Reply:
Used rfind

Reply:
> You can ask for a slice:
> 
> ```
> >>> ''[0]
> ''
> ```

Almost
 ```
>>> ''[:0]
''
```